onnx/defs/schema.cc

```diff
@@ -94,6 +94,63 @@ OpSchemaRegistry* OpSchemaRegistry::Instance() {
   return &instance;
 }
 
+void OpSchema::CheckInputOutputType(struct InferenceContext& ctx) const {
+  std::unordered_map<std::string, std::string> type_constraints;
+  // check all input types
+  for (size_t in_idx = 0; in_idx < ctx.getNumInputs() && in_idx < inputs_.size(); ++in_idx) {
+    const auto& param      = inputs_[in_idx];
+    const auto& type_str   = param.GetTypeStr();
+    const auto& param_type = ctx.getInputType(in_idx);
+    const auto& all_types  = param.GetTypes();
+    if (nullptr == param_type || param_type->value_case() == TypeProto::VALUE_NOT_SET) {
+      continue;
+    } else if (!all_types.empty() && all_types.find(Utils::DataTypeUtils::ToType(*param_type)) == all_types.end()) {
+      fail_check(param.GetName(), " typestr: ", type_str, ", has unsupported type: ", *Utils::DataTypeUtils::ToType(*param_type));
+    }
+    if (param.GetIsHomogeneous()) {
+      const auto& type_proto = Utils::DataTypeUtils::ToType(*param_type);
+      if (type_constraints.find(type_str) == type_constraints.end()) {
+        type_constraints[type_str] = *type_proto;
+      } else if (type_constraints[type_str] != *type_proto) {
+        fail_check(param.GetName(), " has inconsistent type ", *Utils::DataTypeUtils::ToType(*param_type));
+      }
+    }
+  }//for inputs
+  //check all output types
+  for (size_t out_idx = 0; out_idx < ctx.getNumOutputs() && out_idx < outputs_.size(); ++out_idx) {
+    const auto& param      = outputs_[out_idx];
+    const auto& type_str   = param.GetTypeStr();
+    const auto& param_type = ctx.getOutputType(out_idx);
+    const auto& all_types  = param.GetTypes();
+    bool output_type_found = true;
+    //infer type if necessary
+    if (param_type->value_case() == TypeProto::VALUE_NOT_SET) {
+      if (all_types.size() == 1) {
+        *param_type = Utils::DataTypeUtils::ToTypeProto(*all_types.begin());
+      } else if (type_constraints.find(type_str) != type_constraints.end()) {
+        auto data_type = Utils::DataTypeUtils::ToType(type_constraints[type_str]);
+        *param_type = Utils::DataTypeUtils::ToTypeProto(data_type);
+      } else {
+        output_type_found = false;
+      } 
+    }
```

Comment:
The logic looks correct. But I think it can be simplified a bit since if we take the THEN branch above (VALUE_NOT_SET), we set the param_type (to a correct type) and so we don't really need to do the checks below. So, I think I would just make the checking logic below the ELSE part. Does that make sense? The THEN part does inference, and the ELSE part checks correctness of pre-specified type.

Reply:
Yes that's sensible approach - the reason I am not making if an "ELSE" is because if sometime in the future we place some new logic there inferring the type (apart from using those we already have like allowed types and types constraints), we might well wish the inferred type get scrutinized.